code/modules/crafting/items.dm

```diff
@@ -979,7 +979,253 @@
 				/obj/item/weldingtool/advanced,
 				/obj/item/stock_parts/cell/ammo/mfc,
 				/obj/item/stock_parts/cell/ammo/ec,
+				/obj/item/stock_parts/cell/ammo/ecp,
 				/obj/item/megaphone)
 
+/obj/item/experimental
+	name = "Experimental component"
+	desc = "What could this do..."
+	icon = 'icons/obj/assemblies.dmi'
+	icon_state = "radio-multitool"
+
+/obj/item/experimental/attackby(obj/item/W, mob/user, params)
+	. = ..()
+	if(istype(W, /obj/item/gun/ballistic/shotgun))
+		to_chat(usr, "You can't improve [W.name]...")
+		return
+	if(istype(W, /obj/item/gun/ballistic))
+		gun(W, user)
+	if(istype(W, /obj/item/gun/energy))
+		egun(W, user)
+	if(istype(W, /obj/item/clothing/suit/armor))
+		armor(W, user)
+	if(istype(W, /obj/item/clothing/head))
+		hat(W, user)
+	//if(istype(W, /obj/item/clothing/suit/armor/f13/power_armor))
+	//	parmor(W)
+
+/obj/item/experimental/proc/gun(obj/item/W, mob/user)
+	var/obj/item/gun/ballistic/B = W 
+
+	var/dmgmod = rand(-10,10)
+	var/penmod = rand(-10,10)
+	var/spdmod = rand(-10,10)
+	var/overall = dmgmod+penmod-spdmod
+	var/prefix
+
+	if(HAS_TRAIT(user,TRAIT_MASTER_GUNSMITH))
+		dmgmod += 2
+		penmod += 2
+		spdmod += 2
+		overall = dmgmod+penmod-spdmod
+
+	if(B.tinkered > 0 && !HAS_TRAIT(user,TRAIT_MASTER_GUNSMITH))
+		to_chat(usr, "You have already tinkered with this item.")
+		return
+
+	if(B.tinkered > 2 && HAS_TRAIT(user,TRAIT_MASTER_GUNSMITH))
+		to_chat(usr, "You have already tinkered with this item too much.")
+		return
+
+	switch(overall)
+		if(-30 to -15)
+			prefix = "Ruined "
+		if(-15 to 0)
+			prefix = "Inferior "
+		if(0 to 15)
+			prefix = "Improved "
+		if(15 to 30)
+			prefix = "Superior "
+		if(30 to 100)
+			prefix = "Legendary "
+	
+	B.extra_damage += (dmgmod)
+	B.extra_penetration += (penmod/30)
+	B.fire_delay += (spdmod/5)
+	B.name = prefix + B.name
+	B.tinkered += 1
+
+	to_chat(usr, "You tinker with the gun making [W.name]...")
```

Comment:
This entire system is fucked, but:
- Use `user` instead of `usr`.
- Cache the old name and say something like `"You tinker with \the [old_name], making it \a [W]."`
- Use a span class, like `SPAN_NOTICE("message here")`. Should probably have it depend on the prefix.